pkg/adapter/v2/cloudevents_test.go

```diff
@@ -293,6 +301,124 @@ func TestNewCloudEventsClient_request(t *testing.T) {
 	}
 }
 
+func TestNewCloudEventsClient_receiver(t *testing.T) {
+	sampleEvent := func() *cloudevents.Event {
+		event := cloudevents.NewEvent(cloudevents.VersionV1)
+		event.SetID("abc-123")
+		event.SetSource("unit/test")
+		event.SetType("unit.type")
+		event.SetDataContentType("application/json")
+		return &event
+	}
+
+	testCases := map[string]struct {
+		Headers      nethttp.Header
+		Body         []byte
+		ExpectedBody string
+		WantTimeout  bool
+	}{
+		"binary event": {
+			Headers: map[string][]string{
+				"content-type":   {"application/json"},
+				"ce-specversion": {"1.0"},
+				"ce-id":          {"abc-123"},
+				"ce-type":        {"unit.type"},
+				"ce-source":      {"unit/test"},
+			},
+			Body:         []byte(`{"type": "binary"}`),
+			ExpectedBody: `{"type": "binary"}`,
+		},
+		"structured event": {
+			Headers: map[string][]string{
+				"content-type": {"application/cloudevents+json"},
+			},
+			Body: []byte(`{
+				"specversion": "1.0",
+				"id": "abc-123",
+				"source": "unit/test",
+				"type": "unit.type",
+				"datacontenttype": "application/json",
+				"data": {"type": "structured"}
+			}`),
+			ExpectedBody: ` {"type": "structured"}`,
+		},
+		"malformed event": {
+			Headers: map[string][]string{
+				"ce-type": {"err.event"},
+			},
+			WantTimeout: true,
+		},
+	}
+	for n, tc := range testCases {
+		t.Run(n, func(t *testing.T) {
+			port := rand.Intn(16383) + 49152 // IANA Dynamic Ports range
+			ceClient, err := NewCloudEventsClientWithOptions(nil, &mockReporter{}, cloudevents.WithPort(port))
+			if err != nil {
+				t.Errorf("failed to create CE client: %v", err)
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+
+			events := make(chan event.Event)
+			defer close(events)
+
+			go func() {
+				err := ceClient.StartReceiver(ctx, func(event event.Event) error {
+					go func() {
+						events <- event
+					}()
+					return nil
+				})
+				if err != nil {
+					t.Errorf("failed to start receiver: %v", err)
+				}
+			}()
+			time.Sleep(5 * time.Millisecond)
```

Comment:
a `WaitGroup` and the sleep duration increased should make this test less racy since as of now is likely that the client sends an event before the receiver is started. 
```suggestion
			events := make(chan event.Event, 1)
			defer close(events)
                        
                      var wg sync.WaitGroup
                      wg.Add(1)
			go func() {
			        wg.Done()
				err := ceClient.StartReceiver(ctx, func(event event.Event) error {
					events <- event
					return nil
				})
				if err != nil {
					t.Errorf("failed to start receiver: %v", err)
				}
			}()
			wg.Wait() // wait for the above goroutine to be running
			time.Sleep(50 * time.Millisecond) // wait for the receiver to start
```

Reply:
thanks for the comments @pierDipi, either the waitgroup or increased sleep duration did the thing!